subprojects/core/src/integTest/groovy/org/gradle/api/internal/project/LifecycleAwareProjectIntegrationTest.groovy

```diff
@@ -87,4 +87,39 @@ class LifecycleAwareProjectIntegrationTest extends AbstractIntegrationSpec {
         then:
         outputContains "a foo=bar1"
     }
+
+    def 'LifecycleAwareProject wrappers are symmetrically equals to wrapped projects and to each other'() {
+        given:
+        settingsFile """
+            include(":a")
+        """
+
+        buildFile """
+            ext["wrappedBySubprojects"] = subprojects.toList().get(0)
+            ext["wrappedByAllprojects"] = allprojects.toList().get(1)
+            ext["wrappedByProject"] = project(":a")
+        """
+
+        buildFile("a/build.gradle", """
+                def wrappedBySubprojectsAccess = rootProject.ext["wrappedBySubprojects"]
+                def wrappedByAllprojectsAccess = rootProject.ext["wrappedByAllprojects"]
+                def wrappedByProjectAccess = rootProject.ext["wrappedByProject"]
+
+                def projects = [wrappedByAllprojectsAccess, wrappedBySubprojectsAccess, wrappedByProjectAccess, project]
+
+                [projects, projects].combinations()
+                    .findAll { a, b -> a.class != b.class }
```

Comment:
I'm a bit wary about conditional logic in the build-under-test build logic.
As an idea, we can accumulate projects in a `Map<Project, List<Project>>` from all wrappers and print that map, asserting its state. Alternatively, it can be a counting map -- `Map<Project, Integer>`.

Reply:
Moved to a task and a build service in the new test